sstables/mx/partition_reversing_data_source.cc

```diff
@@ -0,0 +1,608 @@
+// true.
+// After reading a tombstone, current_tombstone_reversing_info() returns
+// information about the tombstone kind, as well as the offsets of its
+// members, which is useful for reversing the tombstone.
+//
+// `row_body_skipping_context` does not handle the static row (if there is one in the partition),
+// only `unfiltered`s (clustering rows and tombstones).
+class row_body_skipping_context : public data_consumer::continuous_data_consumer<row_body_skipping_context> {
+    bool _end_of_partition = false;
+    bool _finished = false;
+    processing_result_generator _gen;
+    temporary_buffer<char>* _processing_data;
+
+public:
+    struct tombstone_reversing_info {
+        uint64_t kind_offset;
+        bound_kind_m range_tombstone_kind;
+
+        // Range tombstone markers in the sstable data file come in two kinds: bound markers and boundary markers.
+        // Bound markers happen when a range tombstone opens or ends.
+        // Boundary markers happen when one range tombstone ends but another opens at the same position.
+        //
+        // Bound markers have one `delta_deletion_time` structs (tombstone timestamp + local deletion time) at the end.
+        // Boundary markers have two.
+        //
+        // `first_deletion_time_offset` gives the position of the first `delta_deletion_time` (which is present for both kinds),
+        // after_first_deletion_time gives its end position (i.e. position of last byte plus one), which in case of boundary
+        // markers is the start position of the second `delta_deletion_time` (in case of bound markers its the end of the whole marker).
+        uint64_t first_deletion_time_offset;
+        uint64_t after_first_deletion_time_offset;
+    };
+private:
+    unfiltered_flags_m _flags{0};
+    std::optional<tombstone_reversing_info> _current_tombstone_reversing_info;
+    uint64_t _prev_unfiltered_size;
+
+    // for calculating the clustering blocks
+    boost::iterator_range<std::vector<std::optional<uint32_t>>::const_iterator> _ck_column_value_fix_lengths;
+    uint64_t _ck_blocks_header;
+    uint32_t _ck_blocks_header_offset;
+    bool _reading_range_tombstone_ck = false;
+    bool _reading_partition_ck = false;
+    uint16_t _ck_size;
+    column_translation _column_translation;
+
+    void setup_ck(const std::vector<std::optional<uint32_t>>& column_value_fix_lengths) {
+        if (column_value_fix_lengths.empty()) {
+            _ck_column_value_fix_lengths = boost::make_iterator_range(column_value_fix_lengths);
+        } else {
+            _ck_column_value_fix_lengths = boost::make_iterator_range(std::begin(column_value_fix_lengths),
+                                                                      std::begin(column_value_fix_lengths) + _ck_size);
+        }
+        _ck_blocks_header_offset = 0u;
+    }
+    bool no_more_ck_blocks() const { return _ck_column_value_fix_lengths.empty(); }
+    void move_to_next_ck_block() {
```

Comment:
Not able to say after a quick glance, would have to do some deeper research. I'd suggest to leave nontrivial changes such as this for a followup considering this PR already went through 10 versions and we're already late with estimated delivery time, and it would also require modifying the code that @wmitros prepared for backporting to enterprise (a version of this code which does not use coroutines)
(also, the mx reader code which this code is based off of also does not use `clustering_parser`)

Reply:
We can leave this for a follow up.